h4l/selection/lepton.py

```diff
@@ -0,0 +1,126 @@
+            if self.config_inst.campaign.x.preUL : # pre-UL WP for Run II (miniAOD branch: Run2_CutBased_BTag16)
+                # print("This is preUL!")
+                # from IPython import embed; embed()
+                return ((pt<=10.) &     ((fSCeta<0.8                   & BDT > 0.85216885148) | \
+                                        (fSCeta>=0.8 & fSCeta<1.479 & BDT > 0.82684550976) | \
+                                        (fSCeta>=1.479                & BDT > 0.86937630022))) \
+                        | ((pt>10.) &  ((fSCeta<0.8                   & BDT > 0.98248928759) | \
+                                        (fSCeta>=0.8 & fSCeta<1.479 & BDT > 0.96919224579) | \
+                                        (fSCeta>=1.479                & BDT > 0.79349796445)))
+
+            else: # UL WP (miniAOD branch Run2_CutBased_UL)
+                # print("This is not preUL!")
+                # from IPython import embed; embed()
+                return ((pt<=10.) &     (((fSCeta<0.8)                   & (BDT > 0.9128577458)) | \
+                                        ((fSCeta>=0.8) & (fSCeta<1.479) & (BDT > 0.9056792368)) | \
+                                        ((fSCeta>=1.479)                & (BDT > 0.9439440575)))) \
+                        | ((pt>10.) &  (((fSCeta<0.8)                   & (BDT > 0.1559788054)) | \
+                                        ((fSCeta>=0.8) & (fSCeta<1.479) & (BDT > 0.0273863727)) | \
+                                        ((fSCeta>=1.479)                & (BDT > -0.5532483665))))
+        return ak.ones_like(events.Electron.pt)
+    default_mask = (
+        (pt > min_pt) &
+        (abs(events.Electron.eta) < 2.5) & 
+        (events.Electron.dxy < 0.5) &
+        (events.Electron.dz < 1.0) & 
+        (abs(events.Electron.sip3d) < 4) & 
+        return_cuts()
+    )
+    sorted_idx = ak.argsort(events.Electron.pt, axis=1, ascending=False)
+    # filter unselected electron indices
+    selected_electron_idx = sorted_idx[default_mask[sorted_idx]]
+    
+
+
+    return events, SelectionResult(
+        objects = {
+            "Electron": {
+                "Electron": selected_electron_idx
+            }
+        }
+    )
+
+@selector(
+    uses=({
+        f"Muon.{var}" for var in [
+            # four momenta information
+            "pt", "eta", "phi", "mass", 
+            # quality criteria
+            "isGlobal", "isStandalone", "isTracker", "nStations", "nTrackerLayers",
+            # impact parameters
+            "dxy", "dz", "sip3d",
+            # IDs
+            "tightId", "mvaId", "highPtId",
+            #isolation
+            "pfRelIso03_all",
+        ]}
+    ),
+    exposed=False,
+)
+def muon_selector(self: Selector, events: ak.Array, **kwargs) -> tuple[ak.Array, SelectionResult]:
+    min_pt = 15
+    max_eta = 2.4
+    selected_muon_mask = (
+        # Global or Tracker Muon
+        (events.Muon.isGlobal | events.Muon.isTracker) &
```

Comment:
in the 4l analysis we require:

`
mu.isGlobal or (mu.isTracker and mu.nStations>0)
`
it should not be a big deal for the didactic purpose of the exercise, but I wanted to point this out, as it should be a minimal change, maybe it can be implemented